internal/pkg/vm/state/state.go

```diff
@@ -0,0 +1,204 @@
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/encoding"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"github.com/pkg/errors"
+)
+
+// Review: can we get rid of this?
+type Tree interface {
+	Root() (Root, bool)
+
+	SetActor(ctx context.Context, key actorKey, a *actor.Actor) error
+	GetActor(ctx context.Context, key actorKey) (*actor.Actor, bool, error)
+	DeleteActor(ctx context.Context, key actorKey) error
+
+	Commit(ctx context.Context) (Root, error)
+
+	GetAllActors(ctx context.Context) <-chan GetAllActorsResult
+}
+
+// TreeBitWidth is the bit width of the HAMT used to store a state tree
+const TreeBitWidth = 5
+
+// Root is the root type of a state.
+//
+// Any and all state can be identified by this.
+//
+// Note: it might not be possible to locally reconstruct the entire state if the some parts are missing.
+type Root = cid.Cid
+
+type actorKey = address.Address
+
+// State is the VM state manager.
+type State struct {
+	store    cbor.IpldStore
+	dirty    bool
+	root     Root
+	rootNode *hamt.Node
+}
+
+// NewState creates a new VM state.
+func NewState(store cbor.IpldStore) *State {
+	// Review: we should probably start the tree as dirty given that the current root does not match the state it will persist
```

Comment:
I think this is only used for testing and genesis creation. An alternative would be for this method to write the empty hamt root to the store, the side-effect explicit with a name like `CreateEmptyState()`.

But otherwise, yes dirty.